File: ponio/examples/curtiss_hirschfelder_exprk.py
# Copyright 2022 PONIO TEAM. All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

#!/usr/bin/env python

import subprocess
import os
import argparse

import matplotlib.pyplot as plt
import matplotlib.cm as cm
import numpy as np

name = "curtiss_hirschfelder_exprk"
data_dir = name+"_data"
img_dir = data_dir

parser = argparse.ArgumentParser(
    description=f"Compile, launch and plot results of example `{name}`")
parser.add_argument('--only-save', action='store_true',
                    help="Just save output in img directory")

arguments = parser.parse_args()

if arguments.only_save:
    img_dir = os.path.join("img", name)
    os.makedirs(img_dir, exist_ok=True)


make = subprocess.Popen(["make", name])
make.wait()

args = [os.path.join(".", name)]
process = subprocess.Popen(args)
process.wait()

methods = {
    "rk44": "RK(4, 4)",
    "krogstad": "expRK(4, 4) Krogstad",
    "lrk44": "LRK(4, 4)"
}

plt.rcParams["figure.figsize"] = (8, 4)

line_style = "-"
for tag, meth in methods.items():
    data = np.loadtxt(os.path.join(data_dir, f"{tag}.dat"))
    t, y, dt = data[:, 0], data[:, 1], data[:, 2]

    plt.plot(t, y, line_style, label=meth)
    line_style = "--"

plt.ylabel("$y$")
plt.xlabel('time')

plt.legend()

plt.savefig(os.path.join(img_dir, "01.png"), dpi=200)

if not arguments.only_save:
    plt.title("Solution of Curtiss Hirschfelder problem")
    plt.show()

File: ponio/examples/curtiss_hirschfelder_exprk.cpp
// Copyright 2022 PONIO TEAM. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// NOLINTBEGIN(misc-include-cleaner)

#include <cmath>
#include <filesystem>
#include <string>

#include <ponio/observer.hpp>
#include <ponio/problem.hpp>
#include <ponio/runge_kutta.hpp>
#include <ponio/solver.hpp>
#include <ponio/time_span.hpp>

// NOLINTEND(misc-include-cleaner)

/*
solve Curtiss and Hirschfelder problem:

$$
    \begin{aligned}
        \dot{y} =  k(\cos(t) - y) \\
        y(0) = y_0
    \end{aligned}
$$

$y_0 = 2$
*/

int
main()
{
    std::string const dirname = "curtiss_hirschfelder_exprk_data";

    using state_t = double;

    double const tf = 2.0;
    double const dt = 0.05;

    double const k = 50;

    auto linear_part    = -k;
    auto nonlinear_part = [=]( double t, double )
    {
        return k * std::cos( t );
    };

    auto pb_curtiss_hirshfelder = ponio::make_lawson_problem( linear_part, nonlinear_part );

    state_t const y_0                    = 2.0;
    ponio::time_span<double> const tspan = { 0., tf };

    { // test with RK(4, 4) method
        auto filename = std::filesystem::path( dirname ) / "rk44.dat";
        auto obs      = ponio::observer::file_observer( filename );

        ponio::solve( pb_curtiss_hirshfelder, ponio::runge_kutta::rk_44(), y_0, tspan, dt, obs );
    }

    { // test with expRK Krogstad method
        auto filename = std::filesystem::path( dirname ) / "krogstad.dat";
        auto obs      = ponio::observer::file_observer( filename );

        ponio::solve( pb_curtiss_hirshfelder, ponio::runge_kutta::krogstad(), y_0, tspan, dt, obs );
    }

    { // test with LRK(4, 4) method
        auto filename = std::filesystem::path( dirname ) / "lrk44.dat";
        auto obs      = ponio::observer::file_observer( filename );

        auto exp = []( double x )
        {
            return std::exp( x );
        };

        ponio::solve( pb_curtiss_hirshfelder, ponio::runge_kutta::lrk_44( exp ), y_0, tspan, dt, obs );
    }

    return 0;
}
